packages/upload-client/src/blob.js

```diff
@@ -0,0 +1,398 @@
+      cause: result.out.error,
+    })
+  }
+
+  const nextTasks = parseBlobAddReceiptNext(result)
+
+  const { receipt } = nextTasks.allocate
+  /* c8 ignore next 5 */
+  if (!receipt.out.ok) {
+    throw new Error(`failed ${BlobCapabilities.add.can} invocation`, {
+      cause: receipt.out.error,
+    })
+  }
+
+  const { address } = receipt.out.ok
+  if (address) {
+    const fetchWithUploadProgress =
+      options.fetchWithUploadProgress ||
+      options.fetch ||
+      globalThis.fetch.bind(globalThis)
+
+    let fetchDidCallUploadProgressCb = false
+    const { status } = await retry(
+      async () => {
+        try {
+          const res = await fetchWithUploadProgress(address.url, {
+            method: 'PUT',
+            mode: 'cors',
+            body: bytes,
+            headers: address.headers,
+            signal: options.signal,
+            onUploadProgress: (status) => {
+              fetchDidCallUploadProgressCb = true
+              if (options.onUploadProgress)
+                createUploadProgressHandler(
+                  address.url,
+                  options.onUploadProgress
+                )(status)
+            },
+            // @ts-expect-error - this is needed by recent versions of node - see https://github.com/bluesky-social/atproto/pull/470 for more info
+            duplex: 'half',
+          })
+          if (res.status >= 400 && res.status < 500) {
+            throw new AbortError(`upload failed: ${res.status}`)
+          }
+          return res
+        } catch (err) {
+          if (options.signal?.aborted === true) {
+            throw new AbortError('upload aborted')
+          }
+          throw err
+        }
+      },
+      {
+        retries: options.retries ?? REQUEST_RETRIES,
+      }
+    )
+    if (status !== 200) throw new Error(`upload failed: ${status}`)
+
+    if (!fetchDidCallUploadProgressCb && options.onUploadProgress) {
+      // the fetch implementation didn't support onUploadProgress
+      const blob = new Blob([bytes])
+      options.onUploadProgress({
+        total: blob.size,
+        loaded: blob.size,
+        lengthComputable: false,
+      })
+    }
+  }
+
+  // Invoke `conclude` with `http/put` receipt
+  const derivedSigner = ed25519.from(
+    /** @type {import('@ucanto/interface').SignerArchive<import('@ucanto/interface').DID, typeof ed25519.signatureCode>} */
+    (nextTasks.put.task.facts[0]['keys'])
+  )
+
+  const httpPutReceipt = await Receipt.issue({
+    issuer: derivedSigner,
+    ran: nextTasks.put.task.cid,
+    result: { ok: {} },
+  })
+  const httpPutConcludeInvocation = createConcludeInvocation(
+    issuer,
+    // @ts-expect-error object of type unknown
+    audience,
+    httpPutReceipt
+  )
+  const ucanConclude = await httpPutConcludeInvocation.execute(conn)
+
+  if (!ucanConclude.out.ok) {
+    throw new Error(`failed ${BlobCapabilities.add.can} invocation`, {
+      cause: result.out.error,
+    })
+  }
+
+  return multihash
```

Comment:
I would expect blob/add not to return until task corresponding to `result.out.ok.site` is done, which I realize from the current state of the backend implementation have no good way of doing. Never the less I'll try to make a PR that will add missing pieces. I don't think we should block this in the meantime.